### PR TITLE
fix(i2_s): remove unsafe to_float cast + guard BLAS I2_S routing

### LINKED ISSUE #468: UB in to_float callback for I2_S (missing scale param) + 3 more bugs on ARM Ampere Altra
## Environment

- **Platform**: Oracle Cloud Compute A1 Flex (Ampere Altra, armv8.2-a, `asimddp` present in `/proc/cpuinfo`)
- **OS**: Ubuntu 22.04 LTS
- **Compiler**: clang 14.0
- **Commit**: main branch, March 2026
- **Build flags**:
  ```
  -DBITNET_ARM_TL1=ON
  -DCMAKE_C_COMPILER=clang
  -DCMAKE_CXX_COMPILER=clang++
  -DCMAKE_C_FLAGS="-march=armv8.2-a+fp16+dotprod"
  ```
- **Model**: `microsoft/BitNet-b1.58-2B-4T-gguf` (`ggml-model-i2_s.gguf`, 1.2GB)
- **Python**: 3.12 (venv)

## System info (from llama-cli output)
```
NEON = 1 | ARM_FMA = 1 | SVE = 0 | MATMUL_INT8 = 0 | (all x86 = 0)
```

---

## Problem

With `-b 1` (gemv path): model produces real words but **completely incoherent** output.
With `-b 32`+ (gemm path for prompt): model produces **GGGG garbage** for all generated tokens.

```bash
# gemv path (-b 1, default in run_inference.py)
$ python run_inference.py -m models/BitNet-b1.58-2B-4T-gguf/ggml-model-i2_s.gguf \
    -p "The capital of France is" -n 30 -t 4

The capital of France is offline officially.ai happy collective lower sl ruled m unit proven-for...

# gemm path (-b 32)
$ ./build/bin/llama-cli -m models/BitNet-b1.58-2B-4T-gguf/ggml-model-i2_s.gguf \
    -p "Hello, my name is" -n 50 -t 4 -b 32

Hello, my name isGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG
```

Speed is correct (~18 tok/s on 4 ARM cores), model loads without errors.

---

## Bug 1: `to_float` callback — type mismatch (confirmed, patched)

In `3rdparty/llama.cpp/ggml/src/ggml.c`, the type table for `GGML_TYPE_I2_S` registers `dequantize_row_i2_s` as a `ggml_to_float_t` callback, but the signatures do not match:

```c
// Typedef (3 params):
typedef void (*ggml_to_float_t)(const void * x, float * y, int64_t n);

// Actual function (4 params — 4th is the i2 scale):
void dequantize_row_i2_s(const uint8_t * x, float * y, int64_t n, const float i2_scale);

// Bug in type table (ggml.c ~line 1185):
.to_float = (ggml_to_float_t) dequantize_row_i2_s,  // scale never passed → UB
```

The cast silences the compiler but causes undefined behavior — on ARM, the 4th argument register is garbage → `i2_scale` is random → dequantized weights are garbage.

**Applied fix** — wrapper in the same translation unit:
```c
// At top of ggml.c:
static void dequantize_row_i2_s_wrapper(const void * vx, float * y, int64_t n) {
    const uint8_t * x = (const uint8_t *) vx;
    const float i2_scale = ((const float *)(x + (n / 4)))[0];
    dequantize_row_i2_s(x, y, n, i2_scale);
}

// In type table:
.to_float = dequantize_row_i2_s_wrapper,
```

**Effect**: without this fix → GGGG for all configs. With fix → real words with `-b 1` (but still incoherent, see Bug 3).

---

## Bug 2: `+dotprod` not enabled in CMakeLists for ARM (build issue)

The build system generates `-march=armv8.2-a` without `+dotprod`. On ARM Ampere Altra:
- CPU has `asimddp` (dotprod SDOT/UDOT) in `/proc/cpuinfo`
- Without `+dotprod`, `__ARM_FEATURE_DOTPROD` is not defined by clang
- `ggml-bitnet-mad.cpp` falls to the `#else` path using `vmlal_s8` (int16 accumulator)
- For vectors of 2560+ elements, int16 accumulator overflows → incorrect results (related to #411)

**Verification**:
```bash
clang -march=armv8.2-a+fp16+dotprod -dM -E - < /dev/null | grep DOTPROD
# → #define __ARM_FEATURE_DOTPROD 1

clang -march=armv8.2-a+fp16 -dM -E - < /dev/null | grep DOTPROD
# → (nothing)
```

**Suggested fix**: add `+dotprod` to CMakeLists.txt for AArch64 targets when CPU supports it, or document it as a required flag.

---

## Bug 3: gemm path (`ne11 > 3`) produces GGGG on ARM — root cause unknown

When the batch has more than 3 tokens (prompt evaluation with `-b 32`), the code takes the `ggml_gemm_i2_i8_s` path instead of `ggml_gemv_i2_i8_s`. On ARM, this path produces completely wrong output:

```
# Relevant code path in ggml.c (~line 13268):
if (gemm && (ne11 > 3)) {
    if (src0->type == GGML_TYPE_I2_S) {
        // gemm path — produces GGGG on ARM
        const float * scale = (float *)((uint8_t*)(src0->data) + (ne00 * ne01 / 4));
        gemm(...);
        for (...) {
            tmp[...] = (tmp[...] - act_sums[col]) / (act_scales[col]) * (*scale);
        }
    }
}
```

Since prompt tokens are processed via gemm and corrupt the KV cache, all subsequent generated tokens are also garbage.

**Workaround**: use `-b 1` to force all computation through the gemv path (avoids the bug but is slower and still produces incoherent text).

---

## Bug 4: incoherent output even with gemv path (remaining issue)

Even with `-b 1` + Bug 1 fix + dotprod enabled, output is semantically incoherent. The model generates real vocabulary tokens but with no semantic coherence:

```
"The capital of France is offline officially.ai happy collective..."
"Hello, my name is Freedom reduce Dispatch linear Dahl tent corps..."
```

This persists even with temperature=0 (greedy decoding), suggesting the issue is in the logit distribution rather than sampling. Possible causes:
- `act_sums` (activation zero-point correction) computed incorrectly for ARM
- Scale offset wrong for GQA tensors (K/V: [2560, 640] with 5 KV heads vs Q: [2560, 2560])
- Activation quantization (`quantize_row_i8_s`) producing incorrect scale values

---

## Steps to reproduce

```bash
# 1. Clone and build (Ubuntu 22.04, ARM Ampere)
git clone https://github.com/microsoft/BitNet.git && cd BitNet
git submodule update --init --depth=1
python3 -m venv .venv && source .venv/bin/activate && pip install -r requirements.txt

# 2. Build (add +dotprod — not in default CMakeLists)
cmake -S . -B build \
  -DBITNET_ARM_TL1=ON \
  -DCMAKE_C_COMPILER=clang \
  -DCMAKE_CXX_COMPILER=clang++ \
  -DCMAKE_C_FLAGS="-march=armv8.2-a+fp16+dotprod" \
  -DCMAKE_CXX_FLAGS="-march=armv8.2-a+fp16+dotprod" \
  -DCMAKE_BUILD_TYPE=Release
make -C build -j4

# 3. Download model
huggingface-cli download microsoft/BitNet-b1.58-2B-4T-gguf \
  --local-dir models/BitNet-b1.58-2B-4T-gguf

# 4. Test (without Bug 1 fix — GGGG for all batch sizes)
python run_inference.py -m models/BitNet-b1.58-2B-4T-gguf/ggml-model-i2_s.gguf \
  -p "The capital of France is" -n 30

# 5. Apply Bug 1 fix to 3rdparty/llama.cpp/ggml/src/ggml.c (see above)
# Rebuild and test — now produces words with -b 1 but still incoherent
```

---

## Related issues
- #411 — int16 overflow in NEON 1x1 kernel (same root area, different manifestation)
- #195 — ARM inference issues (similar platform)

## Hardware confirmation
Tested exclusively on **Oracle Cloud Compute A1 Flex** (ARM Ampere Altra).
Reports from Reddit/r/LocalLLaMA indicate M1/M2/M3 Mac users achieve coherent output — suggesting the issue may be ARM Ampere-specific or related to the dotprod/vmlal path difference.

### LINKED ISSUE #512: Apple Silicon Metal + BLAS segfault for I2_S when ubatch >= 32 routes generic MUL_MAT
## Environment

- **Platform**: Apple Silicon Mac
- **Host**: Apple M4 Max
- **OS**: macOS
- **Compiler**: Homebrew clang 18.1.8
- **BitNet / submodule state**: BitNet using vendored `3rdparty/llama.cpp` at Eddie-Wang1120/llama.cpp commit `1f86f058de0c3f4098dedae2ae8653c335c868a1`
- **Model**: `microsoft/BitNet-b1.58-2B-4T-gguf` / `ggml-model-i2_s.gguf`
- **Build flags**:
  - `GGML_METAL=ON`
  - `GGML_ACCELERATE=ON`
  - `GGML_BLAS=ON`
  - `GGML_BLAS_VENDOR=Apple`
  - `BITNET_ARM_TL1=OFF`

## Problem

On Apple Silicon with Metal enabled, `i2_s` inference can segfault when BLAS is enabled and the physical micro-batch crosses the BLAS routing threshold.

The crash is tied to **physical `ubatch`**, not logical `batch`:

- `-b 2048 -ub 31` -> stable
- `-b 32 -ub 31` -> stable
- `-b 2048 -ub 32` -> segfault
- `-b 2048 -ub 512` -> segfault

This means the failure starts exactly when the BLAS backend begins claiming the generic `MUL_MAT` path for larger batches.

## Control Experiment

The same Metal runtime is stable when BLAS is disabled:

- BLAS **ON** + `-b 2048 -ub 512` -> segfault
- BLAS **OFF** + `-b 2048 -ub 512` -> stable

This strongly suggests the crash is in the BLAS-side handling of `GGML_TYPE_I2_S`, not in Metal itself and not in the outer chat request schema.

## Root Cause

`ggml-blas.cpp` allows the generic BLAS `MUL_MAT` path to accept quantized source tensors when `ggml_get_type_traits(src0->type)->to_float != NULL`.

For `GGML_TYPE_I2_S`, that is not safe:

- `I2_S` stores an external scale outside the per-row payload
- the generic BLAS dequantize-to-float path assumes self-contained per-row data
- once `ubatch >= 32`, BLAS starts claiming `MUL_MAT`
- that eventually crashes in the `i2_s` dequant / BLAS matmul path

In crash reports, the top frames consistently land in:

- `dequantize_row_i2_s`
- `ggml_backend_blas_mul_mat`

## Proposed Fix

Reject `GGML_TYPE_I2_S` in the generic BLAS `MUL_MAT` support check so that `I2_S` continues using its specialized non-BLAS path:

```cpp
return src0->type != GGML_TYPE_I2_S &&
       ggml_is_contiguous(src0) &&
       ggml_is_contiguous(src1) &&
       src1->type == GGML_TYPE_F32 &&
       (ne0 >= min_batch && ne1 >= min_batch && ne10 >= min_batch) &&
       (src0->type == GGML_TYPE_F32 || ggml_get_type_traits(src0->type)->to_float != NULL);
```

## Result After Patch

After applying the BLAS guard above:

- BLAS **ON** + Metal + `-b 2048 -ub 512` is stable
- managed broker end-to-end requests no longer segfault under the same settings

This does **not** solve all `i2_s` quality issues, but it does remove the native crash path.

## Related Issues

- #468
- #470
- #195
- #411

### PR DESCRIPTION
**Pull request by JasonOA888**
Fixes #468, Fixes #512

## Bug 1: to_float callback missing scale parameter (UB)

The type traits entry for GGML_TYPE_I2_S registers dequantize_row_i2_s as a ggml_to_float_t callback, but the signatures don't match: the function takes 4 params (x, y, n, i2_scale) but the typedef only has 3. The cast silences the compiler but causes UB — the 4th argument register contains garbage, producing wrong dequantized weights on ARM and potentially x86.

Fix: Add a wrapper function that extracts the scale from the row data (stored after the quantized payload) before calling the real dequantizer.

## Bug 2: BLAS routes I2_S through generic MUL_MAT (segfault on Apple Silicon)

When ubatch >= 32, the BLAS backend claims the generic MUL_MAT path for I2_S because to_float is non-NULL. But I2_S stores an external scale outside the per-row payload, which the generic BLAS dequantize path doesn't handle. This causes segfaults on Apple Silicon Metal+BLAS.

Fix: Reject GGML_TYPE_I2_S from the BLAS MUL_MAT support check so I2_S uses its specialized kernel path.

## Bug 3: llama-server ignores rope freq_base from GGUF metadata

llama-server's update_slots does not pass the model's rope freq_base to the position embedding computation. llama-cli works because it reads the metadata directly. The missing freq_base causes position embeddings to grow unchecked, producing NaN in softmax.

Fix: Ensure the server context reads and applies rope freq_base from model metadata when computing position embeddings.

## Testing

- Verified the wrapper extracts scale correctly (matches the direct-call path in ggml_compute_forward_get_rows_i2_s)
- BLAS guard is a single condition addition, no performance impact
- Server fix aligns with how llama-cli handles freq_base

### CODE REVIEW

**Comment by steev** [comment]
Hi Jason,

Attempting to try your fix, I am not able to, and there are complaints that it doesn't exist.  Is there another fork of llama that this should be pointed at?  Is it something from the upstream llama?

```
~/dev/llm/bitnet$ git submodule update --init --recursive
fatal: remote error: upload-pack: not our ref 6f8a942da0316ccd7e722c057c1fc0586c276b10
fatal: Fetched in submodule path '3rdparty/llama.cpp', but it did not contain 6f8a942da0316ccd7e722c057c1fc0586c276b10. Direct fetching of that commit failed.
```

**Comment by JasonOA888** [comment]
Hi @steev, thanks for flagging this — you're right, the submodule commit `6f8a942` is unreachable because it was on a local llama.cpp build that was never pushed to a public fork.

The root issue is that all three fixes (I2_S to_float UB, BLAS I2_S guard, and rope freq_base) live inside the `3rdparty/llama.cpp` submodule, so the PR only changes the submodule pointer.

I'm going to close this PR and redo it properly:
1. Fork `llama.cpp` and push the fixes there
2. Submit the llama.cpp fixes to the upstream `ggerganov/llama.cpp` (or the BitNet-specific fork `Eddie-Wang1120/llama.cpp`)
3. Then update BitNet's submodule reference

Sorry for the inconvenience.
